### ISSUE TITLE
show-read-articles no should retain the unread article for current feed

### ISSUE THREAD

**Issue by allanwind** [newsboat/newsboat#1715]
**Newsboat version (copy the output of `newsboat -v` or the first line of `git show`)**:
```
newsboat 2.13.0 - https://newsboat.org/
Copyright (C) 2006-2015 Andreas Krennmair
Copyright (C) 2015-2018 Alexander Batischev
Copyright (C) 2006-2017 Newsbeuter contributors
Copyright (C) 2017-2018 Newsboat contributors

Newsboat is free software licensed under the MIT License. (Type `newsboat -vv' to see the full text.)
It bundles JSON for Modern C++ library, licensed under the MIT License: https://github.com/nlohmann/json
It bundles an alphanum algorithm implementation licensed under the MIT license: http://www.davekoelle.com/alphanum.html

newsboat 2.13.0
System: Linux 4.19.0-17-amd64 (x86_64)
Compiler: g++ 8.3.0
ncurses: ncurses 6.1.20181013 (compiled with 6.1)
libcurl: libcurl/7.64.0 GnuTLS/3.6.7 zlib/1.2.11 libidn2/2.0.5 libpsl/0.20.2 (+libidn2/2.0.5) libssh2/1.8.0 nghttp2/1.36.0 librtmp/2.3 (compiled with 7.64.0)
SQLite: 3.27.2 (compiled with 3.27.2)
libxml2: compiled with 2.9.4
```

Config file (copy from ~/.newsboat/config or ~/.config/newsboat/config):

```
articlelist-title-format "%U"
auto-reload yes
bind-key J next-feed
bind-key K prev-feed
bind-key j next
bind-key k prev
datetime-format "%F %T"
error-log ~/.newsboat/error.log
feedlist-format "%4i %n %11u %L - %t"
ignore-article "*" "title =~ \"...\"
ignore-mode display
keep-articles-days 365
reload-threads 32
show-read-articles no
show-read-feeds no
unbind-key C
```

Steps to reproduce the issue:

1. open feed with ENTER
2. open article with ENTER
3. return to index with q and observe that the article you just viewed is no longer shown in article index.  Quite often, I find that I would like to have open it.  The work-around is to toggle-show-read-feeds with l to find the article again.  It would really nice if the now read article would remain in the index till you close the feed.  This would be consistent with how toggle-article-read works N.  (And I do know, that you can open-in-browser with o while viewing the article newsboat).

**Other info you think is relevant**:

Similar behavior for show-read-feeds no would be nice as well.  There is no good work-around for it.

**Comment by Minoru** [comment]
Thanks for filing this! We'll probably have to extend `purge-deleted` (<kbd>$</kbd>) to remove those "stale" articles as well.

**Comment by 097115** [comment]
> This would be consistent with how toggle-article-read works N

While in this open article, you may indeed press `N` instead of `q`, and it will mark the article unread, return you back to the feed, and the article will be present in the list.

**Comment by allanwind** [comment]
I don't realize till I am back in the article index view that I
actually wanted to read that article.  Either upon reflection
or it was a mistake as enter followed by q workflow becomes 
semi-automatic after a while.  You could, and I think that is what 
you are getting at, remap q to toggle-article-read, but most of 
the time the stock behavior is the correct one.

The fundamental problem is that I expect quit on an article to
return me to a previous state other than article marked as read,
but it returns me to a different state where read articles are
hidden.  This is surprising.

It's similar surprising for the feed index, and the work-around is
not as great.  toogle-show-read-feeds, sort by lastupdated, then
browse backwards to see if it was the last before that you just
visited.  This only happen a few times to me, and I gave up rather
than trying to figure out what I was just looking at.

I would call this is a low prior ux issue.

**Comment by allanwind** [comment]
The automatic or manual reload also hides read articles when
in the article list view.  The automatic reload is surprising.

**Comment by 097115** [comment]
> You could, and I think that is what you are getting at, remap q to toggle-article-read

I wouldn't go that far :)

All I'm saying is that there are **two** ways of getting back to the list, with `q` (the article mark read and is no longer shown in the index) and with `N` (article is not mark read, and is still shown in the index). Seems quite logical to me personally (and I'm using the same `show-read-articles no` setup as you).

> q workflow becomes semi-automatic after a while

Well, it really shouldn't. Different tasks require different commands (hotkeys), and we usually don't just hit our keyboards without thinking :)

**Comment by 097115** [comment]
The point is, with `show-read-articles no`, still seeing in the list the article that was already opened and marked as read, is not exactly coherent. Because the whole point for read articles is not to be present in the list any more.

If you find this contradicting to the case when we mark an article as read (with `N`), and it's still visible in the list, then I'd say it's exactly the latter that should be corrected, and not the other way around: marking article as read should immediately hide it from the list, as we specifically told `newsboat` in our preferences not display read articles.

**Comment by allanwind** [comment]
On 2021-07-22T22:31:58, 097115 wrote:
>The point is, with `show-read-articles no`, still seeing in the 
>list the article that was already opened and marked as read, is 
>not exactly coherent. Because the whole point for read articles 
>is not to be present in the list any more.

If you mark an article as read with `N` (toggle-article-read), or
`O` (open-in-browser-and-mark-read), the article remains in the
list marked as read.  This works great.  I am simply asking for 
the same _consistent_ behavior when open the article within 
newsboat (and ideally, similar behavior for the feedlist).  Not 
sure if there is anything else I can say to clarify the request.  
Close the issue if it doesn't make sense.  I don't see the point 
of arguing about it.

**Comment by 097115** [comment]
I'm not a dev, and just showing another perspective on the issue: you think that opening/closing an article acts inconsistent, and I think that it's marking an article read that acts inconsistent.

For @Minoru to decide :)

**Comment by allanwind** [comment]
No worries.  Thanks for the suggestions.

**Comment by Minoru** [comment]
@097115, I like logic, but usability is important too. The problem here is mostly about the inconvenience of undoing an action. One way to fix that would be to have an "undo" keybind; the users will have to learn it then. Another way, the one that @allanwind proposes, is to keep items around, in which case "undo" is just <kbd>N</kbd>. Since the latter is intuitive (users already know <kbd>N</kbd>) and more powerful (one can undo actions in any order), I'm leaning towards that solution.

**Comment by 097115** [comment]
Yeah, except undoing is exactly 1 key in both cases, am I right? :)

Anyway, let's start from the very beginning: if we **explicitly state** in our config files that we don't want to see read items, then why are you planning to shove them through our throat? :) 

I mean, this story has users from both sides. And I personally not only **like** the current behaviour but actually **use i right nowt**, so why are you taking it from from those like me?

**Comment by allanwind** [comment]
> Yeah, except undoing is exactly 1 key in both cases, am I right? :)

@Minoru captured it well with "undo" and I almost wrote it up that way initially, but for me it's not so much toggling the read status as it's about being able to read the title and easily open it  You conveniently gloss over the cases now where read articles are not removed (toggle-article-read, open-in-browser-and-mark-read, mark-feed-read (while in artilistlist)).  There is, afaik, no good work-around for the same issue with the feedlist even with `show-read-feeds yes` and sorted by lastupdate as you can\'t tell what feed you were just operating on irregardless how you sort the list.

> Anyway, let's start from the very beginning: if we **explicitly state** in our config files that we don't want to see read items, then why are you planning to shove them through our throat? :)

It's not specified well enough to make it a particular strong argument.  Here is what the documentation says: "If set to yes, then all articles of a feed are listed in the article list. If set to no, then only unread articles are listed."  As per above, I listed 3 cases where the code behaves differently than what you want it to mean.

> I mean, this story has users from both sides. And I personally not only **like** the current behaviour but I actually **use it right now**, so why are you planning to take it away from those like me?

The **like** part means you have a preference for the current behavior, or alternatively you dislike my suggestion (no offense taken).  That's cool.  It might be helpful if you expressed that more objectively.  For example:

a. It is harder to read the next article title when it's not the first article on the list , or
b. I think it's confusing to how article tiles that I already read, or
c. If the cursor returns on the article you just read, then it will take an extra keystroke to read the next article (currently `enter` vs next + `enter`).

The **use** part is not a valid argument.  Everyone that currently have `show-read-articles no` use the current behavior including me as there is no choice.

We have no idea what other users think of the current or the proposed change.  No one is taking anything away, it's adding something, and granted you may not like what is adding. I don't know how the project make these choices, i.e. whoever implements it, whoever merges it.  For instance it might be a merge condition that change in behavior is behind a configuration key etc.  I might scratch my itch and make the change in a fork, and you may decide to give it a whirl and then have even better arguments (for or against it).

**Comment by 097115** [comment]
Oh, for Christ's sake...

> "undo"

You opened an article, it's marked as read as you open it, and you are **undoing** it with `N`. End of story, no puppies harmed.

> We have no idea

Yeah, pretty much.

Now, it takes exactly two keypresses to open an article, read it and **then immediately hide it**, which is a god send in a feed with a lot of articles. You are suggesting to **complicate** this workflow for really no good reason at all (except that you don't like to press `N` and want to rule everything in the world with `q`).

**Comment by Minoru** [comment]
> Yeah, except undoing is exactly 1 key, am I right? :)

I don't think you are. It takes at least four keypresses to undo an erroneous <kbd>q</kbd> in the articleview: show read articles again with <kbd>l</kbd>, move to the previous one with <kbd>K</kbd>, mark it unread with <kbd>N</kbd>, then hide read articles again with <kbd>l</kbd>. It may take many more in the articlelist, depending on where the previous feed is.

Your suggestion of pressing <kbd>N</kbd> instead of <kbd>q</kbd> is irrelevant because the user has *already* pressed <kbd>q</kbd>. Telling them that they shouldn't have done that is fair, but pointless.

> I personally not only like the current behaviour but I actually use it right now, so why are you planning to take it away from those like me?

If you do use it, please describe a use-case which will be negatively impacted by the proposed change. Use the same structure as in bug reports: I do this, this, and that, the expected result is this, the actual result is that, which is inconvenient because X. Then I can impassively weigh both use-cases and figure out if I want to support one, both, or neither. You are already getting at it in your last comment, but the "because X" part is missing — you only say that it's a "god send in a feed with a lot of articles".

**Comment by 097115** [comment]
> I don't think you are.

Except I am:

> You opened an article, it's marked as read as you open it, and you are undoing it with `N`. End of story, no puppies harmed.

The point is, Allan doesn't want to press `N`. He wants to keep pressing `q` without thinking about consequences.

> If you do use it, please describe a use-case

What exactly is not clear, though?

I open a feed, open some article, read it, close then, and this article is no longer present in the list. A number of pending articles decreases. Exactly as in a todo list. And why is all that? **Because I explicitly stated in my config that I don't want to see read articles.** Exactly according to the manual that some people here didn't bother to read. 

> Then I can impassively weigh

Oh, please. This feature (and it's surely a feature, and not a bug) has been present in Newsbeuter (not even Newsboat then) for ages. It wasn't you, who thought it through, it wasn't you, who implemented it, and thus you definitely should not alter its behaviour.

Crave to help Allan? Add an option `i_m_allan_wide`, which upon setting to true, will open articles with `q`, close articles with `q`, and if pressed and not released, will also make a coffee with `q`.

But ruining an existing workflow, which acts precisely according to the manual, certainly doesn't look like a decision any reasonable person should be making.

**Comment by Minoru** [comment]
It seems we're talking past each other, so let's stop. You failed to convince me, so I'll go with the proposed change.

For the future, please remember that insinuations aren't welcome here. You've been warned.

**Comment by 097115** [comment]
I wasn't trying to "convince" anyone -- was just saying that you are changing this established behaviour only because a person doesn't want to follow a manual.

Literally, the only presented argument in favour of this change was "q workflow becomes semi-automatic after a while" .

How goddamn wise.

**Comment by 097115** [comment]
I wasn't trying to "convince" anyone -- was just saying that you are changing this established behaviour and adding excessive keypresses (**and not even making the change optional, for Christ's sake**) only because a person doesn't want to follow a manual.

Literally, the only presented argument in favour of this change was "q workflow becomes semi-automatic after a while" .

How goddamn wise.
